File: Codeforces/Contest/1043div3/The Cunning Seller (easy version).cpp
#include <bits/stdc++.h>
using namespace std;

void solve() {
    long long n;
    cin >> n;

    long long k = 1;
    long long coins = 0;
    unordered_map<long long, long long> d;
    d[1] = 3;
    long long ans = 0;

    while (pow(3, k) <= n) {
        coins = pow(3, k + 1) + k * pow(3, k - 1);
        d[(long long)pow(3, k)] = d.count((long long)pow(3, k)) ? d[(long long)pow(3, k)] + coins : coins;
        k++;
    }

    if (coins != 0) {
        ans += coins;
        n -= pow(3, k - 1);
    }

    while (n > 1) {
        // cout << "Entered" << endl;
        int p = (int)(log(n) / log(3));
        ans += d[(long long)pow(3, p)];
        n -= pow(3, p);
    }

    if (n == 1) {
        ans += 3;
    }

    cout << ans << "\n";
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    freopen("input.txt", "r", stdin);
    freopen("output.txt", "w", stdout);

    int tc;
    cin >> tc;
    while (tc--) {
        solve();
    }

    return 0;
}

File: Codeforces/Contest/1043div3/The Cunning Seller (easy version).py
import math
import sys

sys.stdin = open('input.txt','r')
sys.stdout = open('output.txt','w')
def exp(base:int,pow:int)->int:
    res = 1
    while pow > 0:
        if pow & 1:
            res *= base
        
        base *= base
        pow >>= 1
    return res
def Log3(n:int)->int:
    val = 1
    p = 0
    while val * 3 <= n:
        val *= 3
        p += 1
    return p
tc = 1
tc = int(input())
def solve():
    n = int(input())
    k = 1
    coins = 0
    d = {1:3}
    ans = 0
    # operations = 0
    while exp(3,k) <= n:
        coins = exp(3,k + 1) + k * exp(3,k - 1)
        d[exp(3,k)] = d.get(exp(3,k) , 0) + coins
        k += 1
    # operation += 1
    # print('d is: ',d)
    if coins != 0:
        ans += coins
        n -= exp(3,k - 1)
    while n > 1:
        # print('Entered')
        pow = Log3(n)
        # print(f'power is: {pow}')
        ans += d[exp(3,pow)]
        n -= exp(3,pow)
    if n == 1:
        ans += 3
    print(ans)
    
    
for _ in range(tc):
    solve()